### ISSUE TITLE
`dataclasses.InitVar` not handled correctly by `sphinx.ext.autodoc`

### ISSUE THREAD

**Issue by mattyclarkson** [sphinx-doc/sphinx#9508]
### Describe the bug

`dataclasses.InitVar` is not correctly identified as a initialisation variable for a `dataclass`

Attempted to document the following:

```
@dataclass
class Bug:
    """A buggy class"""

    i: int
    """A plain old attribute"""

    iv: InitVar[int]
    """An attribute that only is available to the initialisation routine"""

    def __post_init__(self, iv: int) -> None:
        self.i *= iv
```

Will result in:

```
sphinx-build-dataclasses-initvar/bug/dataclass.py:docstring of bug.dataclass.Bug::py:class reference target not found: InitVar
```

When it should certainly resolve to `dataclasses.InitVar`.

### How to Reproduce

```
$ git clone https://github.com/mattyclarkson/sphinx-build-dataclasses-initvar.git
$ cd sphinx-build-dataclasses-initvar
$ sphinx-build \
  -W \
  -j auto \
  -b dirhtml \
  -d sphinx/doctree \
  -c sphinx \
  bug \
  sphinx/html
```

### Expected behavior

The `InitVar` should be correctly resolved. It should _also_ be correctly identified as an intialization only variable and _not_ a class member.

### Your project

https://github.com/mattyclarkson/sphinx-build-dataclasses-initvar

### OS

Arch Linux

### Python version

3.9

### Sphinx version

4.0.0

### Sphinx extensions

sphinx.ext.autodoc, sphinx.ext.intersphinx

**Comment by mattyclarkson** [comment]
FYI: @tk0miya. Seems like you contribute a lot to the `sphinx.ext.autodoc` 😄

**Comment by tk0miya** [comment]
I found `typing.get_type_hints()` raises TypeError for a variable annotated by `dataclasses.InitVar` when lazy annotations evaluation-feature is enabled (a.k.a `__future__.annotations`).

```
$ cat test.py
from __future__ import annotations
from dataclasses import dataclass, InitVar
from typing import get_type_hints

@dataclass
class Foo:
    attr: InitVar[int]


get_type_hints(Foo)
```

```
$ python -V
Python 3.9.6
$ python test.py
Traceback (most recent call last):
  File "/Users/tkomiya/work/sphinx/test.py", line 10, in <module>
    get_type_hints(Foo)
  File "/Users/tkomiya/.pyenv/versions/3.9.6/lib/python3.9/typing.py", line 1424, in get_type_hints
    value = _eval_type(value, base_globals, localns)
  File "/Users/tkomiya/.pyenv/versions/3.9.6/lib/python3.9/typing.py", line 290, in _eval_type
    return t._evaluate(globalns, localns, recursive_guard)
  File "/Users/tkomiya/.pyenv/versions/3.9.6/lib/python3.9/typing.py", line 545, in _evaluate
    type_ =_type_check(
  File "/Users/tkomiya/.pyenv/versions/3.9.6/lib/python3.9/typing.py", line 164, in _type_check
    raise TypeError(f"{msg} Got {arg!r:.100}.")
TypeError: Forward references must evaluate to types. Got dataclasses.InitVar[int].
```

I think this is a bug of Python.

**Comment by tk0miya** [comment]
I posted this case as https://bugs.python.org/issue44799

**Comment by mattyclarkson** [comment]
I assume that `TypeError` is handled [here](https://github.com/sphinx-doc/sphinx/blob/c66b3ad519ed91e79205ee973bfdba17fc06453f/sphinx/util/inspect.py#L651)?

I'm wondering if I can hook into any of the Sphinx events to provide a `get_type_hints` implementation that _does_ understand `InitVar`. i.e. can I capture one of [these](https://github.com/sphinx-doc/sphinx/blob/c66b3ad519ed91e79205ee973bfdba17fc06453f/sphinx/ext/autodoc/__init__.py#L2819-L2823) events and do something hacky for now like regex the `InitVar` part away?

**Comment by mattyclarkson** [comment]
Maybe [PEP649](https://www.python.org/dev/peps/pep-0649) might fix this in a future version and can be `__future__` backported 🤞

**Comment by mattyclarkson** [comment]
Will be fixed in Python 3.11
